qiskit/dagcircuit/dagcircuit.py

```diff
@@ -1986,9 +1986,7 @@ def remove_op_node(self, node):
                 "node type was wrongly provided."
             )
 
-        self._multi_graph.remove_node_retain_edges(
-            node._node_id, use_outgoing=False, condition=lambda edge1, edge2: edge1 == edge2
-        )
+        self._multi_graph.remove_node_retain_edges_by_id(node._node_id)
```

Comment:
I think `by_id` is slightly risky here while bits still have an equality that isn't entirely referential.

I'm fairly sure it's possible to end up with edges that are constructed from different `bit` instances, since the objects that go into the wire payload are passed in by the user to `apply_operation_back`. It sounds unlikely, but I'm fairly sure it is possible, and I'd hate to debug that.

Maybe use `by_key` just until we've closed that loophole?

Reply:
I was thinking about that, but I wasn't too worried because I was thinking the only object identity we care about is between incoming and outgoing edge pairs around a node. We don't have to worry about using the same object globally for a dag as long as the local edge weights use the same object. I was thinking that we'll always use the same object when doing that because we always add the edges in pairs.

But if you think it's safe to use `by_key` I'll update this.

Reply:
I wouldn't put it past either us or a user pass doing `canonical_register = QuantumRegister("q", dag.num_qubits)` and using that in `apply_operation_back` on an existing circuit. That's still technically allowed behaviour, so the key is probably safer for now. The move to Rust space or removal of the register/index bit properties will let us swap to the preferential id one.

Reply:
But that wouldn't matter for this as `apply_operation_back()` uses rustworkx's `insert_node_on_in_edges_multiple()` which pulls the edge weights from the existing edge on the graph and just uses the same object for the two new edges it adds when inserting a node in between an existing edge. That's what I was saying, the only way I could see us using a different object id is if we manipulated the graph directly to do so, otherwise we do everything together.

But as I said before I'll update this to use by_key instead if you're worried about the implications here it's not that much slower.

Reply:
Ahh yeah, I forgot that the weight given isn't used verbatim on the edges in the split, just used to make a key function. In that case I'm fine with id.